src/linux/mod.rs

```diff
@@ -103,49 +111,79 @@ impl<'a> fmt::Debug for SegmentIter<'a> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         let ref phdr = self.inner.as_slice()[0];
 
-        f.debug_struct("SegmentIter").field("phdr", &DebugPhdr(phdr)).finish()
+        f.debug_struct("SegmentIter")
+            .field("phdr", &DebugPhdr(phdr))
+            .finish()
     }
 }
 
 /// A shared library on Linux.
-#[derive(Clone, Copy)]
 pub struct SharedLibrary<'a> {
     size: usize,
     addr: *const u8,
-    name: &'a CStr,
+    name: Cow<'a, CStr>,
     headers: &'a [Phdr],
 }
 
 struct IterState<F> {
     f: F,
     panic: Option<Box<Any + Send>>,
+    idx: usize,
 }
 
 const CONTINUE: libc::c_int = 0;
 const BREAK: libc::c_int = 1;
 
 impl<'a> SharedLibrary<'a> {
-    unsafe fn new(info: &'a libc::dl_phdr_info, size: usize) -> Self {
+    unsafe fn new(info: &'a libc::dl_phdr_info, size: usize, is_first_lib: bool) -> Self {
+        // try to get the name from the dl_phdr_info.  If that fails there are two
+        // cases we can and need to deal with.  The first one is if we are the first
+        // loaded library in which case the name is the executable which we can
+        // discover via env::current_exe (reads the proc/self symlink).
+        //
+        // Otherwise if we have a no name we might be a dylib that was loaded with
+        // dlopen in which case we can use dladdr to recover the name.
+        let mut name = Cow::Borrowed(if info.dlpi_name.is_null() {
+            CStr::from_bytes_with_nul_unchecked(b"\0")
+        } else {
+            CStr::from_ptr(info.dlpi_name)
+        });
+        if name.to_bytes().is_empty() {
+            if is_first_lib {
+                if let Ok(exe) = current_exe() {
+                    name = Cow::Owned(CString::from_vec_unchecked(exe.into_os_string().into_vec()));
+                }
+            } else {
+                let mut dlinfo: libc::Dl_info = mem::zeroed();
+                if libc::dladdr(info.dlpi_addr as *const libc::c_void, &mut dlinfo) != 0 {
```

Comment:
How sure are you that this is useful? I naively would have assumed that `dladdr` derives from the same information returned by `dl_iterate_phdr`.

Reply:
This was suggested by this mongodb issue https://jira.mongodb.org/browse/SERVER-41063